src/quickgui/attributes/qgsquicksubmodel.cpp

```diff
@@ -0,0 +1,183 @@
+{
+  if ( !mModel )
+    return mRootIndex;
+
+  QModelIndex sourceIndex = mModel->index( row, column, parent.isValid() ? mapToSource( parent ) : static_cast<QModelIndex>( mRootIndex ) );
+  return mapFromSource( sourceIndex );
+}
+
+QModelIndex QgsQuickSubModel::parent( const QModelIndex &child ) const
+{
+  if ( !mModel )
+    return mRootIndex;
+
+  QModelIndex idx = mModel->parent( child );
+  if ( idx == mRootIndex )
+    return QModelIndex();
+  else
+    return mapFromSource( idx );
+}
+
+int QgsQuickSubModel::rowCount( const QModelIndex &parent ) const
+{
+  if ( !mModel )
+    return 0;
+
+  return mModel->rowCount( parent.isValid() ? mapToSource( parent ) : static_cast<QModelIndex>( mRootIndex ) );
+}
+
+int QgsQuickSubModel::columnCount( const QModelIndex &parent ) const
+{
+  if ( !mModel )
+    return 0;
+
+  return mModel->columnCount( parent.isValid() ? mapToSource( parent ) : static_cast<QModelIndex>( mRootIndex ) );
+}
+
+QVariant QgsQuickSubModel::data( const QModelIndex &index, int role ) const
+{
+  if ( !mModel )
+    return QVariant();
+
+  return mModel->data( mapToSource( index ), role );
+}
+
+bool QgsQuickSubModel::setData( const QModelIndex &index, const QVariant &value, int role )
+{
+  if ( !mModel )
+    return false;
+
+  return mModel->setData( mapToSource( index ), value, role );
+}
+
+QHash<int, QByteArray> QgsQuickSubModel::roleNames() const
+{
+  if ( !mModel )
+    return QHash<int, QByteArray>();
+
+  return mModel->roleNames();
+}
+
+QModelIndex QgsQuickSubModel::rootIndex() const
+{
+  return mRootIndex;
+}
+
+void QgsQuickSubModel::setRootIndex( const QModelIndex &rootIndex )
+{
+  if ( rootIndex == mRootIndex )
+    return;
+
+  beginResetModel();
+  mRootIndex = rootIndex;
+  endResetModel();
+  emit rootIndexChanged();
+}
+
+QAbstractItemModel *QgsQuickSubModel::model() const
+{
+  return mModel;
+}
+
+void QgsQuickSubModel::setModel( QAbstractItemModel *model )
+{
+  if ( model == mModel )
+    return;
+
+  if ( model )
+  {
+    connect( model, &QAbstractItemModel::rowsAboutToBeInserted, this, &QgsQuickSubModel::onRowsAboutToBeInserted );
+    connect( model, &QAbstractItemModel::rowsInserted, this, &QgsQuickSubModel::onRowsInserted );
+    connect( model, &QAbstractItemModel::rowsAboutToBeRemoved, this, &QgsQuickSubModel::onRowsAboutToBeRemoved );
+    connect( model, &QAbstractItemModel::rowsRemoved, this, &QgsQuickSubModel::onRowsRemoved );
+    connect( model, &QAbstractItemModel::modelAboutToBeReset, this, &QgsQuickSubModel::onModelAboutToBeReset );
+    connect( model, &QAbstractItemModel::modelReset, this, &QAbstractItemModel::modelReset );
+    connect( model, &QAbstractItemModel::dataChanged, this, &QgsQuickSubModel::onDataChanged );
+  }
+
+  mModel = model;
+  emit modelChanged();
+}
+
+void QgsQuickSubModel::onRowsAboutToBeInserted( const QModelIndex &parent, int first, int last )
+{
+  emit beginInsertRows( mapFromSource( parent ), first, last );
```

Comment:
This is not a signal - do not use "emit"

Same below for onRowsInserted, beginRemoveRows, endRemoveRows